[BUG] govc metric.ls panic

**Describe the bug**

When running `govc metric.ls` the command panics

**To Reproduce**
Steps to reproduce the behavior:

1. run `govc metric.ls` against a virtual machine

**Expected behavior**

No panic

**Affected version**
Multiple versions: 0.27.4, 0.28.0

**Screenshots/Debug Output**

```
➜  ~ ./govc metric.ls -debug "$vm"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc939bf]

goroutine 1 [running]:
github.com/vmware/govmomi/performance.groupPerfCounterInfo.Less({0xc0004e8510?, {0xc000222000?, 0xc000543b68?, 0x6d000c0001ee380?}}, 0x4?, 0x3e?)
        /home/runner/work/govmomi/govmomi/performance/manager.go:208 +0x5f
sort.medianOfThree({0x13cae28, 0xc00053e700}, 0x203000?, 0x203000?, 0xc00053e700?)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:89 +0x42
sort.doPivot({0x13cae28, 0xc00053e700}, 0x0, 0xb8)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:114 +0x72
sort.quickSort({0x13cae28, 0xc00053e700}, 0xc0000d5ce8?, 0xc00053e700?, 0xc0000d5ce8?)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:203 +0x6c
sort.Sort({0x13cae28, 0xc00053e700})
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:231 +0x53
github.com/vmware/govmomi/performance.(*Manager).AvailableMetric(0xc0001ee230, {0x13cb050, 0xc00012e000}, {{0xc00014a220?, 0x1?}, {0xc00014a243?, 0xc9a3eb?}}, 0x14)
        /home/runner/work/govmomi/govmomi/performance/manager.go:253 +0x29a
github.com/vmware/govmomi/govc/metric.(*ls).Run(0xc000152560, {0x13cb050, 0xc00012e000}, 0x9?)
        /home/runner/work/govmomi/govmomi/govc/metric/ls.go:178 +0x136
github.com/vmware/govmomi/govc/cli.Run({0xc00012c050, 0x3, 0x3})
        /home/runner/work/govmomi/govmomi/govc/cli/command.go:165 +0x32d
main.main()
        /home/runner/work/govmomi/govmomi/govc/main.go:113 +0x4f
➜  ~
```

**Additional context**

This issue started after our VMC SDDC was upgraded:
SDDC Version: | 1.16v7

https://docs.vmware.com/en/VMware-Cloud-on-AWS/services/com.vmware.vmc-aws-operations/GUID-52CED8FB-2E3A-4766-8C59-2EAD8E2C1D31.html

vCenter:
Version: 7.0.3 Build: 19504857 

We have been running this command in CI for 9 months without issue.
https://github.com/openshift/release/pull/20598/files

Reproduced in non-vmc vcenter as well
Version: 7.0.3 Build: 19234570 

```
root@jump-002-dallas-rh ~]# govc metric.ls /VMware-On-IBM-Cloud-VPC/vm/vcenter
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc939ea]

goroutine 1 [running]:
github.com/vmware/govmomi/performance.groupPerfCounterInfo.Less({0xc000190630?, {0xc0003e2000?, 0x0?, 0x1f9?}}, 0x0?, 0xfd?)
        /home/runner/work/govmomi/govmomi/performance/manager.go:209 +0x8a
sort.doPivot({0x13cae28, 0xc000199960}, 0x0, 0x1fa)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:136 +0x23c
sort.quickSort({0x13cae28, 0xc000199960}, 0xc00043dce8?, 0xc000199960?, 0xc00043dce8?)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:203 +0x6c
sort.Sort({0x13cae28, 0xc000199960})
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:231 +0x53
github.com/vmware/govmomi/performance.(*Manager).AvailableMetric(0xc00037e000, {0x13cb050, 0xc00009e000}, {{0xc000036720?, 0x1?}, {0xc00003673b?, 0xc9a3eb?}}, 0x14)
        /home/runner/work/govmomi/govmomi/performance/manager.go:253 +0x29a
github.com/vmware/govmomi/govc/metric.(*ls).Run(0xc0000ca560, {0x13cb050, 0xc00009e000}, 0x9?)
        /home/runner/work/govmomi/govmomi/govc/metric/ls.go:178 +0x136
github.com/vmware/govmomi/govc/cli.Run({0xc000090070, 0x2, 0x2})
        /home/runner/work/govmomi/govmomi/govc/cli/command.go:165 +0x32d
main.main()
        /home/runner/work/govmomi/govmomi/govc/main.go:113 +0x4f
[root@jump-002-dallas-rh ~]#
```

And then upgraded to u3e
Build: 19717403 

```
[root@jump-002-dallas-rh ~]# govc metric.ls /VMware-On-IBM-Cloud-VPC/vm/vcenter
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc939ea]

goroutine 1 [running]:
github.com/vmware/govmomi/performance.groupPerfCounterInfo.Less({0xc0000f59e0?, {0xc000490000?, 0x0?, 0x1f9?}}, 0x0?, 0xfd?)
        /home/runner/work/govmomi/govmomi/performance/manager.go:209 +0x8a
sort.doPivot({0x13cae28, 0xc0001261e0}, 0x0, 0x1fa)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:136 +0x23c
sort.quickSort({0x13cae28, 0xc0001261e0}, 0xc0003b5ce8?, 0xc0001261e0?, 0xc0003b5ce8?)
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:203 +0x6c
sort.Sort({0x13cae28, 0xc0001261e0})
        /opt/hostedtoolcache/go/1.18.0/x64/src/sort/sort.go:231 +0x53
github.com/vmware/govmomi/performance.(*Manager).AvailableMetric(0xc0001a4460, {0x13cb050, 0xc000036080}, {{0xc0001d6a50?, 0x1?}, {0xc0001d6a6b?, 0xc9a3eb?}}, 0x14)
        /home/runner/work/govmomi/govmomi/performance/manager.go:253 +0x29a
github.com/vmware/govmomi/govc/metric.(*ls).Run(0xc000074580, {0x13cb050, 0xc000036080}, 0x9?)
        /home/runner/work/govmomi/govmomi/govc/metric/ls.go:178 +0x136
github.com/vmware/govmomi/govc/cli.Run({0xc0000300a0, 0x2, 0x2})
        /home/runner/work/govmomi/govmomi/govc/cli/command.go:165 +0x32d
main.main()
        /home/runner/work/govmomi/govmomi/govc/main.go:113 +0x4f
[root@jump-002-dallas-rh ~]# 
```

--- a/govc/metric/info.go
+++ b/govc/metric/info.go
@@ -185,12 +185,13 @@ func (cmd *info) Run(ctx context.Context, f *flag.FlagSet) error {
 			seen := make(map[int32]bool)
 			for i := range all {
 				id := &all[i]
-				if seen[id.CounterId] {
+				info, ok := nc[id.CounterId]
+				if !ok || seen[id.CounterId] {
 					continue
 				}
 				seen[id.CounterId] = true
 
-				names = append(names, nc[id.CounterId].Name())
+				names = append(names, info.Name())
 			}
 		}
 	}

--- a/govc/metric/ls.go
+++ b/govc/metric/ls.go
@@ -95,8 +95,8 @@ func (r *lsResult) Write(w io.Writer) error {
 
 	for _, id := range r.MetricList {
 		if r.cmd.group != "" {
-			info := r.counters[id.CounterId]
-			if info.GroupInfo.GetElementDescription().Label != r.cmd.group {
+			info, ok := r.counters[id.CounterId]
+			if !ok || info.GroupInfo.GetElementDescription().Label != r.cmd.group {
 				continue
 			}
 		}
@@ -116,7 +116,10 @@ func (r *lsResult) Write(w io.Writer) error {
 	}
 
 	for _, id := range res {
-		info := r.counters[id.CounterId]
+		info, ok := r.counters[id.CounterId]
+		if !ok {
+			continue
+		}
 
 		switch {
 		case r.cmd.long:
@@ -147,9 +150,9 @@ func (r *lsResult) MarshalJSON() ([]byte, error) {
 	m := make(map[string]*types.PerfCounterInfo)
 
 	for _, id := range r.MetricList {
-		info := r.counters[id.CounterId]
-
-		m[info.Name()] = info
+		if info, ok := r.counters[id.CounterId]; ok {
+			m[info.Name()] = info
+		}
 	}
 
 	return json.Marshal(m)

--- a/govc/test/metric.bats
+++ b/govc/test/metric.bats
@@ -13,6 +13,7 @@ load test_helper
 
   host=$(govc ls -t HostSystem ./... | head -n 1)
   pool=$(govc ls -t ResourcePool ./... | head -n 1)
+  vm=$(govc ls -t VirtualMachine ./... | head -n 1)
 
   run govc metric.ls "$host"
   assert_success
@@ -21,6 +22,9 @@ load test_helper
   assert_success
 
   run govc metric.ls "$pool"
+  assert_success
+
+  run govc metric.ls "$vm"
   assert_success
 }
 

--- a/performance/manager.go
+++ b/performance/manager.go
@@ -205,10 +205,18 @@ func (d groupPerfCounterInfo) Len() int {
 func (d groupPerfCounterInfo) Less(i, j int) bool {
 	ci := d.ids[i].CounterId
 	cj := d.ids[j].CounterId
-	gi := d.info[ci].GroupInfo.GetElementDescription()
-	gj := d.info[cj].GroupInfo.GetElementDescription()
 
-	return gi.Key < gj.Key
+	giKey := "-"
+	gjKey := "-"
+
+	if gi, ok := d.info[ci]; ok {
+		giKey = gi.GroupInfo.GetElementDescription().Key
+	}
+	if gj, ok := d.info[cj]; ok {
+		gjKey = gj.GroupInfo.GetElementDescription().Key
+	}
+
+	return giKey < gjKey
 }
 
 func (d groupPerfCounterInfo) Swap(i, j int) {
@@ -455,10 +463,14 @@ func (m *Manager) ToMetricSeries(ctx context.Context, series []types.BasePerfEnt
 
 		for j := range s.Value {
 			v := s.Value[j].(*types.PerfMetricIntSeries)
+			info, ok := counters[v.Id.CounterId]
+			if !ok {
+				continue
+			}
 
 			values = append(values, MetricSeries{
-				Name:     counters[v.Id.CounterId].Name(),
-				unit:     counters[v.Id.CounterId].UnitInfo.GetElementDescription().Key,
+				Name:     info.Name(),
+				unit:     info.UnitInfo.GetElementDescription().Key,
 				Instance: v.Id.Instance,
 				Value:    v.Value,
 			})

--- a/simulator/performance_manager.go
+++ b/simulator/performance_manager.go
@@ -81,14 +81,7 @@ func (p *PerformanceManager) QueryPerfCounter(ctx *Context, req *types.QueryPerf
 	body.Res = new(types.QueryPerfCounterResponse)
 	body.Res.Returnval = make([]types.PerfCounterInfo, len(req.CounterId))
 	for i, id := range req.CounterId {
-		if info, ok := p.perfCounterIndex[id]; !ok {
-			body.Fault_ = Fault("", &types.InvalidArgument{
-				InvalidProperty: "CounterId",
-			})
-			return body
-		} else {
-			body.Res.Returnval[i] = info
-		}
+		body.Res.Returnval[i] = p.perfCounterIndex[id]
 	}
 	return body
 }
@@ -135,6 +128,8 @@ func (p *PerformanceManager) buildAvailablePerfMetricsQueryResponse(ids []types.
 			r.Returnval = append(r.Returnval, types.PerfMetricId{CounterId: id.CounterId, Instance: id.Instance})
 		}
 	}
+	// Add a CounterId without a corresponding PerfCounterInfo entry. See issue #2835
+	r.Returnval = append(r.Returnval, types.PerfMetricId{CounterId: 10042})
 	return r
 }
 

